### ISSUE TITLE
rsync 3.2.3 sometimes hang when run from BackupPC

### ISSUE THREAD

**Issue by jonglezb** [WayneD/rsync#256]
We use BackupPC to backup many machines. We updated some machines to debian 11 (which ships rsync 3.2.3) and started seeing rsync hang forever on the clients. It never happened with previous versions: debian 10 ships with rsync 3.1.3 and it was working fine.

When this bug happens, rsync scans the whole filesystem according to strace, and then gets stuck running `select()` forever:

```
strace: Process 4104384 attached                                                                                  
select(1, [0], [], [0], {tv_sec=52, tv_usec=637639}) = 0 (Timeout)                                                
select(1, [0], [], [0], {tv_sec=60, tv_usec=0}) = 0 (Timeout)                                                     
select(1, [0], [], [0], {tv_sec=60, tv_usec=0}) = 0 (Timeout)                                                     
select(1, [0], [], [0], {tv_sec=60, tv_usec=0}) = 0 (Timeout)                                                     
select(1, [0], [], [0], {tv_sec=60, tv_usec=0}) = 0 (Timeout)                                                     
 ```

This bug is non-deterministic: it happens only on some of our debian11 machines, even though they all have the same setup. But when it happens on a machine, it is 100% reproducible on this machine.

For reference, backuppc starts rsync through SSH this way:

```
/usr/bin/ssh -q -x -o StrictHostKeyChecking=no -l backup $CLIENT_HOSTNAME nice -n 19 sudo /usr/bin/rsync --server --sender --numeric-ids --perms --owner --group -D --links --hard-links --times --block-size=2048 --recursive --checksum-seed=32761 . /
```

The `noop_io_until_death()` function looks like a good candidate for such a select-loop, although I haven't verified that rsync was indeed stuck in this function. I looked at what changed this function recently, and found f9bb8f76ee. Reverting this commit on top of 3.2.3 fixed the issue.

This bug has already been reported in Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=969463 (I have just added a comment with basically the same content as here)

**Comment by jonglezb** [comment]
So, both `noop_io_until_death` and f9bb8f7 turned out to be red herrings :/

With the revert of f9bb8f7, it worked a few times, but now it hangs again.

This time I captured a backtrace when `rsync` is stuck in the select-loop:

```
(gdb) bt
#0  0x00007fbf4d034866 in __GI___select (nfds=nfds@entry=1, readfds=readfds@entry=0x7ffc3aa93560, writefds=writefds@entry=0x7ffc3aa93660, exceptfds=exceptfds@entry=0x7ffc3aa935e0, timeout=timeout@entry=0x7ffc3aa93530)
    at ../sysdeps/unix/sysv/linux/select.c:41
#1  0x000055b7e498fa87 in perform_io (needed=needed@entry=4, flags=flags@entry=17) at io.c:741
#2  0x000055b7e499119e in raw_read_buf (buf=0x7ffc3aa93784 "\377\377\377\377", len=4) at io.c:898
#3  0x000055b7e49935eb in read_buf (f=0, buf=buf@entry=0x7ffc3aa93784 "\377\377\377\377", len=len@entry=4) at io.c:1832
#4  0x000055b7e499445f in read_int (f=<optimized out>) at io.c:1722
#5  read_ndx (f=f@entry=0) at io.c:2232
#6  0x000055b7e496eb97 in read_ndx_and_attrs (f_in=f_in@entry=0, f_out=f_out@entry=1, iflag_ptr=iflag_ptr@entry=0x7ffc3aa93898, type_ptr=type_ptr@entry=0x7ffc3aa93897 "\200", buf=buf@entry=0x7ffc3aa94930 "", 
    len_ptr=len_ptr@entry=0x7ffc3aa9389c) at rsync.c:330
#7  0x000055b7e4978391 in send_files (f_in=f_in@entry=0, f_out=f_out@entry=1) at sender.c:216
#8  0x000055b7e498246a in do_server_sender (argv=<optimized out>, argc=<optimized out>, f_out=1, f_in=0) at main.c:953
#9  start_server (f_in=f_in@entry=0, f_out=f_out@entry=1, argc=<optimized out>, argv=0x55b7e52a1d40) at main.c:1251
#10 0x000055b7e496318f in main (argc=<optimized out>, argv=<optimized out>) at main.c:1809
```

Full backtrace with local variables:
```
(gdb) bt full
#0  0x00007fbf4d034866 in __GI___select (nfds=nfds@entry=1, readfds=readfds@entry=0x7ffc3aa93560, writefds=writefds@entry=0x7ffc3aa93660, exceptfds=exceptfds@entry=0x7ffc3aa935e0, timeout=timeout@entry=0x7ffc3aa93530)
    at ../sysdeps/unix/sysv/linux/select.c:41
        resultvar = 18446744073709551102
        sc_ret = <optimized out>
#1  0x000055b7e498fa87 in perform_io (needed=needed@entry=4, flags=flags@entry=17) at io.c:741
        r_fds = {fds_bits = {1, 0 <repeats 15 times>}}
        e_fds = {fds_bits = {1, 0 <repeats 15 times>}}
        w_fds = {fds_bits = {0 <repeats 16 times>}}
        tv = {tv_sec = 21, tv_usec = 879384}
        cnt = <optimized out>
        max_fd = 0
        empty_buf_len = 4
        out = 0x0
        data = <optimized out>
#2  0x000055b7e499119e in raw_read_buf (buf=0x7ffc3aa93784 "\377\377\377\377", len=4) at io.c:898
        pos = 22044
        data = <optimized out>
#3  0x000055b7e49935eb in read_buf (f=0, buf=buf@entry=0x7ffc3aa93784 "\377\377\377\377", len=len@entry=4) at io.c:1832
No locals.
#4  0x000055b7e499445f in read_int (f=<optimized out>) at io.c:1722
        b = "\377\377\377\377"
        num = <optimized out>
        b = {<optimized out>, <optimized out>, <optimized out>, <optimized out>}
        num = <optimized out>
#5  read_ndx (f=f@entry=0) at io.c:2232
        prev_positive = -1
        prev_negative = 1
        prev_ptr = <optimized out>
        num = <optimized out>
        b = "\377\377\377\377"
#6  0x000055b7e496eb97 in read_ndx_and_attrs (f_in=f_in@entry=0, f_out=f_out@entry=1, iflag_ptr=iflag_ptr@entry=0x7ffc3aa93898, type_ptr=type_ptr@entry=0x7ffc3aa93897 "\200", buf=buf@entry=0x7ffc3aa94930 "", 
    len_ptr=len_ptr@entry=0x7ffc3aa9389c) at rsync.c:330
        len = <optimized out>
        iflags = <optimized out>
        flist = <optimized out>
        fnamecmp_type = 128 '\200'
        ndx = <optimized out>
        read_loop = <optimized out>
#7  0x000055b7e4978391 in send_files (f_in=f_in@entry=0, f_out=f_out@entry=1) at sender.c:216
        fd = <optimized out>
        s = <optimized out>
        mbuf = <optimized out>
        st = {st_dev = 51714, st_ino = 822804, st_nlink = 1, st_mode = 33204, st_uid = 0, st_gid = 43, __pad0 = 0, st_rdev = 0, st_size = 18566016, st_blksize = 4096, st_blocks = 36320, st_atim = {tv_sec = 1635721201, 
            tv_nsec = 141796664}, st_mtim = {tv_sec = 1637160013, tv_nsec = 142838386}, st_ctim = {tv_sec = 1637160013, tv_nsec = 142838386}, __glibc_reserved = {0, 0, 0}}
        fname = "var/log/wtmp\000log\000hoenix.yaml\000ploy.2021-11-17T15:09:33+01:00.1.log\000ophia.grid5000.fr.log\000\060b2a-000000000152ade0-0005d0fc1e1b290d.journal\000\000\355A", '\000' <repeats 23 times>, "\020\000\000\000\000\000\000\000\020\000\000\000\000\000\000\b\000\000\000\000\000\000\000\267\065\202a\001", '\000' <repeats 11 times>...
        xname = "\000\000c\000postgresql-common\000user_clusters\000ions\000Terminal.css\000s\000\000mon\000ion\000xul\000\000\000\000m\000m\000m\000b\000eb\000b\000\061\065.pem\000m\000m\000\060\060\061\065\061b2f4-0005d0faf1b3c1db.journal\000nal\000alic.woff\000\000f\000\000ff\000on-39.pyc\000n-39.pyc\000ml\000.yml\000thon-39.pyc\000ml\000-39.py"...
        path = <optimized out>
        slash = <optimized out>
--Type <RET> for more, q to quit, c to continue without paging--c
        fnamecmp_type = 128 '\200'
        iflags = 98304
        xlen = -1
        file = <optimized out>
        phase = 1
        max_phase = 1
        itemizing = 0
        log_code = FINFO
        f_xfer = 1
        save_io_error = 0
        ndx = <optimized out>
        j = <optimized out>
#8  0x000055b7e498246a in do_server_sender (argv=<optimized out>, argc=<optimized out>, f_out=1, f_in=0) at main.c:953
        flist = <optimized out>
        dir = <optimized out>
        flist = <optimized out>
        dir = <optimized out>
#9  start_server (f_in=f_in@entry=0, f_out=f_out@entry=1, argc=<optimized out>, argv=0x55b7e52a1d40) at main.c:1251
No locals.
#10 0x000055b7e496318f in main (argc=<optimized out>, argv=<optimized out>) at main.c:1809
        ret = <optimized out>
        sigmask = {__val = {33638915, 0 <repeats 15 times>}}
```

**Comment by jonglezb** [comment]
From what I can tell, rsync is trying to read 4 bytes from the other side of the SSH connection (backuppc process), but never receives anything.

According to https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=969463 BackupPC implements the rsync protocol version 28 by itself, so it's possible that a slight change of protocol behaviour in rsync may cause this hanging issue.

**Comment by jonglezb** [comment]
This is actually a duplicate of #95 

See also https://github.com/backuppc/backuppc/issues/369

TL;DR: a change in rsync 3.2.3 exposed an old bug in BackupPC 3.X
